### PR TITLE
cgen: allow assignment operators for type aliases

### PR DESCRIPTION
**Pull request by Delta456**
As per the discussion operator overloading for alias with parent types as primitives are now disallowed as it cause a lot of confusion.

The test is in `vlib/math/big/big_test.v`.

### CODE REVIEW

File: vlib/v/checker/tests/method_op_alias_err.out [hunk 1]
```diff
@@ -0,0 +1,28 @@
+vlib/v/checker/tests/method_op_alias_err.vv:4:1: error: both sides of an operator must be the same type
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/method_op_alias_err.out:1]: more like
```
    4 | fn (f Foo) + (f1 Foo2) Foo2 {
      |                  ~~~~
```
paramater has to be the same type like the receive on overloading.

File: vlib/v/checker/tests/method_op_alias_err.out [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    2 | type Foo2 = string
+    3 | 
+    4 | fn (f Foo) + (f1 Foo2) Foo2 {
+      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    5 |     return Foo2(f + f1)
+    6 | }
+vlib/v/checker/tests/method_op_alias_err.vv:5:19: error: infix expr: cannot use `string` (right expression) as `string`
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/method_op_alias_err.out:8]: this error is fail!

Comment (danieldaeschle) [reply vlib/v/checker/tests/method_op_alias_err.out:8]: it should be something like; `recusive function call` or something. @JalonSolov what do you think?

File: vlib/v/checker/tests/method_op_alias_err.out [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    6 | }
+    7 |
+vlib/v/checker/tests/method_op_alias_err.vv:10:6: error: mismatched types `Foo` and `string`
+    8 | fn main() {
+    9 |    mut f := Foo('fg')
+   10 |    f += 'fg'
+      |      ~~
+   11 |    f -= Foo('fo') 
+   12 |     println(f)
+vlib/v/checker/tests/method_op_alias_err.vv:11:6: error: operation `Foo` - `Foo` does not exist, please define it
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/method_op_alias_err.out:22]: remove the "pelase define it"

**Review by danieldaeschle — CHANGES_REQUESTED** [review]

File: vlib/v/checker/tests/method_op_alias_err.out (hunk 3 above)
Comment (Delta456) [reply vlib/v/checker/tests/method_op_alias_err.out:22]: This will be resolved in next PR.

File: vlib/v/checker/tests/method_op_alias_err.out (hunk 1 above)
Comment (JalonSolov) [reply vlib/v/checker/tests/method_op_alias_err.out:1]: Hmm...  I would use

`operation only supported on similar types`

and I would use that on all operations, no matter the type.  "similar" instead of "same" because regular `+` should work without a problem on `1 + 2.0` - even though they are different types, they are still "similar" (as they are both numbers).

For `+=` the result just has to be castable to the receiver.

File: vlib/v/checker/tests/method_op_alias_err.out (hunk 2 above)
Comment (JalonSolov) [reply vlib/v/checker/tests/method_op_alias_err.out:8]: Again, I don't see this in his changes.  This one definitely needs to be fixed, but in a separate PR.

Comment (Delta456) [reply vlib/v/checker/tests/method_op_alias_err.out:8]: Yes

File: vlib/v/checker/tests/method_op_alias_err.out (hunk 1 above)
Comment (Delta456) [reply vlib/v/checker/tests/method_op_alias_err.out:1]: Yes

**Comment by Delta456** [comment]
Ready for review.

**Comment by Delta456** [comment]
Ready for review.

File: vlib/v/gen/cgen.v [hunk 4]
```diff
@@ -2925,40 +2928,31 @@ fn (mut g Gen) infix_expr(node ast.InfixExpr) {
 	right_sym := g.table.get_type_symbol(node.right_type)
 	has_eq_overloaded := !left_sym.has_method('==')
 	has_ne_overloaded := !left_sym.has_method('!=')
+	has_string_alias_op := ((left_sym.has_method('+') || left_sym.has_method('>') || left_sym.has_method('<') ||
```
Comment (danieldaeschle) [inline vlib/v/gen/cgen.v:2931]: what does "has_string_alias_op" mean?

what happens if symbol has method "+" but method "<" is used?

File: vlib/v/gen/str.v [hunk 5]
```diff
@@ -367,6 +367,7 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype table.Type) {
 	} else {
 		str_fn_name := g.gen_str_for_type(typ)
 		g.write('${str_fn_name}(')
+		println(str_fn_name)
```
Comment (danieldaeschle) [inline vlib/v/gen/str.v:370]: remove this

**Review by danieldaeschle — CHANGES_REQUESTED** [review]
could you add more tests to test edge cases?

File: vlib/v/gen/cgen.v (hunk 4 above)
Comment (Delta456) [reply vlib/v/gen/cgen.v:2931]: `what happens if symbol has method "+" but method "<" is used?` it will then use `string__lt` function.

`what does "has_string_alias_op" mean?` I was thinking to name it `has_string_op_overloaded` but forgot to do it.

File: vlib/v/gen/str.v (hunk 5 above)
Comment (Delta456) [reply vlib/v/gen/str.v:370]: Ok

File: vlib/v/gen/cgen.v (hunk 4 above)
Comment (danieldaeschle) [reply vlib/v/gen/cgen.v:2931]: what about `is_string_op_overloaded`?

File: vlib/v/gen/cgen.v [hunk 6]
```diff
@@ -1920,6 +1920,15 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
...
 			mut str_add := false
 			mut op_overloaded := false
+			mut alias_op_overloading := false
+			extracted_op := match assign_stmt.op {
+					.plus_assign { '+' }
+					.minus_assign { '-' }
+					.div_assign { '/' }
+					.mod_assign { '%' }
+					.mult_assign { '*' }
+					else { 'unknown op' }
```
Comment (danieldaeschle) [inline vlib/v/gen/cgen.v:1930]: could this happen actually?

Comment (danieldaeschle) [reply vlib/v/gen/cgen.v:1930]: maybe add a comment

File: vlib/v/checker/checker.v [hunk 7]
```diff
@@ -2557,11 +2558,22 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
...
 				}
 			} else {
-				if left_name == right_name {
-					c.error('operation `$left_name` $extracted_op `$right_name` does not exist, please define it',
-						assign_stmt.pos)
-				} else {
-					c.error('mismatched types `$left_name` and `$right_name`', assign_stmt.pos)
+				if !(extracted_op == '+' && left_type.is_string()) && (parent_type.is_number() || !parent_type.is_number()) {
+					if left_name == right_name {
+						println('right : $right_name')
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:2563]: remove that println

File: vlib/v/checker/checker.v [hunk 8]
```diff
@@ -2557,11 +2558,22 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
...
+						c.error('operation `$left_name` $extracted_op `$right_name` does not exist, please define it',
+							assign_stmt.pos)
+					} else {
+						c.error('mismatched types `$left_name` and `$right_name`', assign_stmt.pos)
+					}
+				} else if !parent_type.is_number() && extracted_op == '+' && left_type.is_string() {
+					if left_name == right_name {
+						println('left : $left_name')
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:2571]: remove

**Review by danieldaeschle — CHANGES_REQUESTED** [review]

**Comment by Delta456** [comment]
Ready for review.

File: vlib/v/checker/checker.v [hunk 9]
```diff
@@ -2537,7 +2537,13 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
 		}
 		if assign_stmt.op in
 			[.plus_assign, .minus_assign, .mod_assign, .mult_assign, .div_assign] &&
-			left_sym.kind == .struct_ && right_sym.kind == .struct_ {
+			((left_sym.kind == .struct_ && right_sym.kind == .struct_) || (left_sym.kind == .alias)) {
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:2540]: no need to have parentheses around `(left_sym.kind == .alias)`

Comment (danieldaeschle) [reply vlib/v/checker/checker.v:2540]: or is it vfmt? then raise an issue please :)

**Review by danieldaeschle — CHANGES_REQUESTED** [review]

File: vlib/v/checker/checker.v (hunk 9 above)
Comment (Delta456) [reply vlib/v/checker/checker.v:2540]: cc @serkonda7

Comment (serkonda7) [reply vlib/v/checker/checker.v:2540]: I will put it on my list

Comment (Delta456) [reply vlib/v/checker/checker.v]: No need to, rebasing fixes it. I don't know how 😕

**Review by danieldaeschle — APPROVED** [review]

**Comment by Delta456** [comment]
Changes in error messages and `token.Position` suggested by @danieldaeschle will be done in my next PR when this is merged.
